### ISSUE TITLE
We should have a table layout widget

### ISSUE THREAD

**Issue by abarth** [flutter/flutter#1048]
For the particular use case I had, I wanted a fixed number of columns all the same size.  I used a bunch of Rows and Columns with Flexible around each item, but we should have a table layout widget to help.  We might want to just add some more knobs to Grid to make it work.

**Comment by Hixie** [comment]
The main reason I haven't done anything here yet is I'm not sure what the API should look like. I'm thinking maybe something like:

```dart
var t = new RenderTable();
t.cells[2][2] = someRenderBox;
...
t.columns[1].width = 200.0;
t.columns[1].decoration = someBoxDecoration;
...
t.rows[1].height = 100.0;
t.rows[1].decoration = someBoxDecoration;
```

I'm not sure how to do merged cells, though. Maybe if multiple cells have the same RenderBox then the box is drawn over the bounding box of all the cells that the RenderBox is in?

**Comment by abarth** [comment]
I was imagining something more like grid layout from CSS.

```dart
var t = new RenderTable();
t.columnSpec = [ new TableSpec.fixed(50), new TableSpec.flex(1), new TableSpec.fixed(50) ];
t.rowSpec = [ new TableSpec.fixed(50, repeating: true) ]
...
t.children = [ a, b, c, d, e ]
a.parentData
  ..row = 0
  ..col = 0
b.parentData
  ..row = 0
  ..col = 1
  ..colSpan = 2
...
```

The children just paint in order and positioned according to their parent data.

**Comment by Hixie** [comment]
Hm, I didn't think about having to have repeating rows...

**Comment by abarth** [comment]
Here's the one-off widget I built for my use case:

```
final Container _kPlaceholder = new Container(width: 0.0, height: 0.0);

class FlexTable extends StatelessComponent {
  FlexTable({
    Key key,
    this.columnCount,
    this.children
  }) : super(key: key);

  final int columnCount;
  final List<Widget> children;

  Widget build(BuildContext context) {
    List<Widget> rows = <Widget>[];
    int rowCount = (children.length / columnCount).ceil();
    for (int rowIndex = 0; rowIndex < rowCount; ++rowIndex) {
      List<Widget> cols = <Widget>[];
      for (int colIndex = 0; colIndex < columnCount; ++colIndex) {
        int i = colIndex + rowIndex * columnCount;
        cols.add(new Flexible(
          child: i < children.length ? children[i] : _kPlaceholder
        ));
      }
      rows.add(new Row(cols));
    }
    return new Column(rows);
  }
}
```

**Comment by abarth** [comment]
> Hm, I didn't think about having to have repeating rows...

Looks like CSS might not support that:

https://css-tricks.com/snippets/css/complete-guide-grid/

Another wrinkle is that we could allow rows and cols after the repeating point and use negative row/col indices to address them.  Also, you might want to repeat a large sequence:

```
rowSpec = [ new TableSpec.fixed(50.0), new TableSpec.repeat([ new TableSpec.fixed(10.0), new TableSpec.fixed(20.0) ], new TableSpec.fixed(50.0)];
```

Then you could use `col == -1` to get to that last col.  It gets complicated if you have `col == -5`.  You need to walk the children to figure out how many repeats you need.

Ok, I'm starting to think that we shouldn't handle repeating.  We can always do some repeating in the Widget layer instead.  I'm imaging that we'd build several widgets on top of this render object that would handle simpler cases with less verbosity.

**Comment by abarth** [comment]
Ok, I'm going to start out by generalizing our existing RenderGrid to handle more regular grid use cases.
